### PR TITLE
Bundle ios dependencies

### PR DESCRIPTION
**Pull request by christopherfujino**
## Description

This updates the flutter tool cache to download binary files for ideviceinstaller, ios-deploy, libimobiledevice, and dynamically linked dependencies from Flutter's GCP bucket (`https://storage.googleapis.com/flutter_infra/ios-usb-dependencies/$name/$version/$name.zip`)

Version hashes are tracked in `flutter/bin/internal/*.version`.

```
// New files that get cached to flutter/bin/cache/artifacts/
.
├── ideviceinstaller
│   ├── COPYING
│   └── ideviceinstaller
├── ios-deploy
│   ├── LICENSE
│   ├── LICENSE2
│   └── ios-deploy
├── libimobiledevice
│   ├── COPYING
│   ├── COPYING.LESSER
│   ├── libtasn1-LICENSE
│   ├── libtasn1-AUTHORS
│   ├── idevice_id
│   ├── ideviceinfo
│   ├── idevicename
│   ├── idevicescreenshot
│   ├── idevicesyslog
│   └── libimobiledevice.6.dylib
├── libplist
│   ├── COPYING
│   └── libplist.3.dylib
├── openssl
│   ├── LICENSE
│   ├── libcrypto.1.0.0.dylib
│   └── libssl.1.0.0.dylib
└── usbmuxd
    ├── COPYING
    ├── iproxy
    └── libusbmuxd.4.dylib
```

## Testing

Start with a Mac already setup for ios/flutter development. `flutter run` an example app from master on an iOS device, to rule out other setup problems :).

Next, run the following command to unlink these ios/usb dependencies from `$PATH`:

```
$ brew unlink ideviceinstaller ios-deploy libimobiledevice libplist usbmuxd
```

Now on master branch, `flutter doctor` should fail the ios workflow validations. Check out this PR branch:

```
$ git fetch upstream pull/34669/head:pr/34669 && git checkout pr/34669
```

Now re-run doctor, and you should see:

```
~/git/flutter$ flutter doctor
Building flutter tool...
Downloading libimobiledevice...                                     0.4s
Downloading usbmuxd...                                              0.2s
Downloading libplist...                                             0.2s
Downloading ideviceinstaller...                                     0.2s
Downloading ios-deploy...                                           0.2s
Downloading openssl...                                              0.4s
Doctor summary (to see all details, run flutter doctor -v):
```

The tool downloads the artifacts to the cache. Now, to verify the artifacts work, run an app with a iOS device attached.

If the app successfully runs, then the tool is successfully using the cached binaries.

To restore your dev environment to use the homebrew versions:

```
$ brew link ideviceinstaller ios-deploy libimobiledevice libplist usbmuxd
```

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/precache.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/precache.dart]: Should be handled automatically via the `DevelopmentArtifact.iOS`

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/commands/precache.dart]: Cool

**Comment by sjindel-google** [comment]
I tried to test it, but the first command failed:

```
> brew unlink ideviceinstaller ios-deploy libimobiledevice libplist usbmuxd
Unlinking /usr/local/Cellar/ideviceinstaller/1.1.0_4... Error: Permission denied @ unlink_internal - /usr/local/bin/ideviceinstaller
```

When I run it as root, it doesn't work either:

```
> sudo brew unlink ideviceinstaller ios-deploy libimobiledevice libplist usbmuxd
Error: Running Homebrew as root is extremely dangerous and no longer supported.
As Homebrew does not drop privileges on installation you would be giving all
build scripts full access to your system.
```

**Comment by christopherfujino** [comment]
> There's a difference in the launching step (for launching a debug-mode app).

Thank you @sjindel-google, there was a bug which I pushed a fix for. Could you try it again?

**Comment by sjindel-google** [comment]
It works now.

File: packages/flutter_tools/lib/src/cache.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/cache.dart]: Is there any way to handle all the ios usb artifacts together in this class rather than having a subclass-per-file?

File: packages/flutter_tools/lib/src/ios/devices.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/ios/devices.dart]: Is this check still needed for `ideviceinstaller` since we know that it should be in the artifact cache? If this check is not needed, then there's no need to clobber `PATH` when calling `which`, which I wouldn't be surprised to see confuse it.

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/ios/devices.dart]: Good point.

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/ios/devices.dart]: Done

File: packages/flutter_tools/lib/src/cache.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/cache.dart]: Done

File: packages/flutter_tools/lib/src/cache.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/cache.dart]: Add here:
```dart
static const List<String> artifactNames = <String>[
    'libimobiledevice',
    'usbmuxd',
    'libplist',
    'openssl',
    'ideviceinstaller',
    'libtasn1',
    'ios-deploy',
  ];
```

And above use this list as `IosUsbArtifacts.artifactNames`

File: packages/flutter_tools/lib/src/cache.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/cache.dart]: Is there any way to make this a bit less platform specific. For example, we might run into a similar situation with artifacts on Linux. Maybe change this to `artifactExecutionEnv` and have it return a `Map<String, String>`. It would iterate over `_artifacts`, and ask each of them if it wants to add anything to the environment. Then you'd cache the result so it doesn't have to be recomputed each time.

File: packages/flutter_tools/lib/src/ios/mac.dart
Comment (chingjun) [inline packages/flutter_tools/lib/src/ios/mac.dart]: Leave a comment here for tracking. As we talked earlier, it would be easier to replace the tools internally if we:
1. Use an `enum` instead of a `String` to find the tools
2. Temporarily fallback to use the tools in `$PATH` if file is not found, to avoid breaking existing users internally during the transition. After the transition we could remove the fallback.

File: packages/flutter_tools/lib/src/artifacts.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/artifacts.dart]: nit: Instead of repeating `_artifactToFileName(artifact)`, how about assigning to a variable above the switch.

File: packages/flutter_tools/lib/src/artifacts.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/artifacts.dart]: ditto comment w.r.t. `_artifactToFileName(artifact)`

File: packages/flutter_tools/lib/src/cache.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/cache.dart]: Can this be a static const of `IosUsbArtifacts`?

File: packages/flutter_tools/lib/src/cache.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/cache.dart]: Add a comment here like: 'On Mac, any entries that should be added to DYLD_LIB_PATH that are needed to use the artifact.

File: packages/flutter_tools/lib/src/ios/mac.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/ios/mac.dart]: Cache this in a private instance field instead of spawning a process each time.

File: packages/flutter_tools/lib/src/ios/mac.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/ios/mac.dart]: ditto comment for isWorking.

File: packages/flutter_tools/lib/src/artifacts.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/artifacts.dart]: done

File: packages/flutter_tools/lib/src/artifacts.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/artifacts.dart]: done

File: packages/flutter_tools/lib/src/cache.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/cache.dart]: done

File: packages/flutter_tools/lib/src/cache.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/cache.dart]: DONE

File: packages/flutter_tools/lib/src/ios/mac.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/ios/mac.dart]: Done

File: packages/flutter_tools/lib/src/ios/mac.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/ios/mac.dart]: Done

**Comment by jonahwilliams** [comment]
If you pull master in you should get the (now correctly functioning) coverage bot

File: packages/flutter_tools/lib/src/cache.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/cache.dart]: done

File: dev/devicelab/lib/framework/adb.dart
Comment (christopherfujino) [reply dev/devicelab/lib/framework/adb.dart]: Done

File: dev/devicelab/lib/framework/adb.dart
Comment (christopherfujino) [reply dev/devicelab/lib/framework/adb.dart]: Done

File: dev/devicelab/lib/framework/adb.dart
Comment (christopherfujino) [reply dev/devicelab/lib/framework/adb.dart]: done

File: packages/flutter_tools/lib/src/cache.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/cache.dart]: done

File: packages/flutter_tools/lib/src/cache.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/cache.dart]: Yes, it's called in `CachedArtifact`'s `update()` method.

File: packages/flutter_tools/lib/src/ios/mac.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/ios/mac.dart]: done

File: dev/devicelab/lib/framework/adb.dart
Comment (chingjun) [inline dev/devicelab/lib/framework/adb.dart]: Use `artifacts.getArtifactPath` for this one as well.

**Review by chingjun — APPROVED** [review]
LGTM for internal compatibility except for a minor change commented above

File: packages/flutter_tools/lib/src/ios/mac.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/ios/mac.dart]: done

File: dev/devicelab/lib/framework/adb.dart
Comment (christopherfujino) [reply dev/devicelab/lib/framework/adb.dart]: I would prefer to keep the string literal; this string appears throughout multiple files, the most general variable name I could think would be `kMacLinkLibraryPath`, and this seems unlikely to change. And if we were to change it to some other type of path, I think we'd have to change the variable name anyway.

File: dev/devicelab/lib/framework/adb.dart
Comment (christopherfujino) [reply dev/devicelab/lib/framework/adb.dart]: I don't think I can, as this is the devicelab project, and it has no access to the flutter_tools code. Is that right @jonahwilliams?

Comment (jonahwilliams) [reply dev/devicelab/lib/framework/adb.dart]: This package doesn't currently depend on the flutter tool. We could add a dependency here, but that would still require a `src` import. I'm okay with using the hardcoded path since this is for our own internal testing framework

**Comment by christopherfujino** [comment]
> If you pull master in you should get the (now correctly functioning) coverage bot

I did so, however no comment from codecov. Maybe it only reads the `.codecov.yml` when PRs are created, and doesn't check again on updates if it wasn't there initially?

**Comment by jonahwilliams** [comment]
Codecov is broken until https://github.com/flutter/flutter/pull/35657 lands :(

File: dev/devicelab/lib/framework/adb.dart
Comment (chingjun) [reply dev/devicelab/lib/framework/adb.dart]: Oh I didn't realize this is in the test code. In that case I'm fine with that.

**Comment by jonahwilliams** [comment]
The coverage fix has landed

File: packages/flutter_tools/lib/src/ios/devices.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/ios/devices.dart]: nit: long line.

File: packages/flutter_tools/lib/src/ios/devices.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/ios/devices.dart]: ditto

File: packages/flutter_tools/lib/src/ios/devices.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/ios/devices.dart]: ditto

File: packages/flutter_tools/lib/src/ios/devices.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/ios/devices.dart]: return here, then else not needed.

File: packages/flutter_tools/lib/src/ios/mac.dart
Comment (zanderso) [inline packages/flutter_tools/lib/src/ios/mac.dart]: long lines.

File: dev/devicelab/lib/framework/adb.dart
Comment (zanderso) [reply dev/devicelab/lib/framework/adb.dart]: `Cache.dyLdLibPath` could return `MapEntry<String,String>('DYLD_LIBRARY_PATH', ...)`.

**Review by zanderso — APPROVED** [review]
lgtm w/ nits and cleanup suggestion for DYLD_LIBRARY_PATH

File: packages/flutter_tools/lib/src/cache.dart
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/cache.dart]: Note: this should be DevelopmentArtifact.universal, and documented that it is required since we cannot detect that an iOS device is connected without these artifacts

File: dev/devicelab/lib/framework/adb.dart
Comment (christopherfujino) [reply dev/devicelab/lib/framework/adb.dart]: yeah. if we added other env variables later we would have to merge the maps, but that's probably the best solution. i initially wrote it that way (returning a map), changed it to return a string for reasons, but that code is gone now, so i'll change it back.

Comment (christopherfujino) [reply dev/devicelab/lib/framework/adb.dart]: Cool, I didn't know about `MapEntry`!

File: bin/internal/openssl.version
Comment (Hixie) [inline bin/internal/openssl.version]: would it make sense to statically link the libraries?

File: packages/flutter_tools/lib/src/ios/devices.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/ios/devices.dart]: cool

File: packages/flutter_tools/lib/src/artifacts.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/artifacts.dart]: done

File: packages/flutter_tools/lib/src/ios/devices.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/ios/devices.dart]: done

**Review by tvolkert — COMMENTED** [review]
LGTM, though you may want to wait for a more authoritative review before you merge.

Keep https://github.com/flutter/flutter/wiki/Tree-hygiene#squashing-commits in mind when you merge since there are a number of commits to be squashed 🙂

File: packages/flutter_tools/lib/src/ios/devices.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/ios/devices.dart]: I chatted with @jonahwilliams about this, we it doesn't look like it gets instantiated, but we're not sure. Since it could potentially be a breaking change, I created [this issue for it](https://github.com/flutter/flutter/issues/36008) and will land after this PR.

**Comment by christopherfujino** [comment]
> LGTM, though you may want to wait for a more authoritative review before you merge.
> 
> Keep https://github.com/flutter/flutter/wiki/Tree-hygiene#squashing-commits in mind when you merge since there are a number of commits to be squashed 🙂

Thanks, and yes, messy commit messages is one of my pet peeve nits :)

**Pull request merged**
